<a href="https://colab.research.google.com/github/ChojnackiKRZ/machine_learning_kurs/blob/main/supervised/02_regression/01_linear_regression_normal_equation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Przygotowanie danych

In [1]:
import numpy as np

In [2]:
X1 = np.array([1, 2, 3, 4, 5, 6])
Y = np.array([3000, 3250, 3500, 3750, 4000, 4250])
m = len(X1)

print(f'Lata pracy: {X1}')
print(f'Wynagrodzenie: {Y}')
print(f'Liczba próbek: {m}')

Lata pracy: [1 2 3 4 5 6]
Wynagrodzenie: [3000 3250 3500 3750 4000 4250]
Liczba próbek: 6


In [32]:
Y.shape

(6,)

In [3]:
X1 = X1.reshape(m, 1)

In [4]:
bias = np.ones((m, 1))

In [5]:
X = np.append(bias, X1, axis = 1)

Rozwiązanie równania regresji linowej w sklearn

In [36]:
from sklearn.linear_model import LinearRegression

regression = LinearRegression()
regression.fit(X1, Y)

LinearRegression()

In [37]:
print (regression.intercept_)
print (regression.coef_)

2750.0
[250.]


Końcowa postać modelu$$Y = 2750 + 250X_1$$

Rozwiązanie równiania regresji liniowej przy użyciu równania normalnego

### <a name='2'></a> Równanie normalne

Regresja liniowa w $R^2$:
$$Y = w_0 + w_1X_1$$
$$Y = W^TX$$
gdzie: $$W= \left[\begin{matrix}w_0\\w_1\end{matrix}\right]$$ $$ X= \left[\begin{matrix}1\\X_1\end{matrix}\right] $$ stąd $$ W^T= \left[\begin{matrix}w_0&w_1\end{matrix}\right] $$    
$$Y = W^TX = \left[\begin{matrix}w_0&w_1\end{matrix}\right] \cdot  \left[\begin{matrix}1\\X_1\end{matrix}\right] = w_0 + w_1X_1$$

Równanie normalne - równanie pozwalające obliczyć minimum funkcji straty (o ile istnieje)
$$W = (X^TX)^{-1}(X^TY)$$

In [6]:
first = np.linalg.inv(np.dot(X.T, X))
first

array([[ 0.86666667, -0.2       ],
       [-0.2       ,  0.05714286]])

In [9]:
second = np.dot(X.T, Y)
second

array([21750., 80500.])

In [10]:
W = np.dot(first, second)
W

array([2750.,  250.])

Końcowa postać modelu$$Y = 2750 + 250X_1$$